### PR TITLE
invensense/icm42688p: use full 20 bit data, increase ODR, disable all filters

### PR DESCRIPTION
**Pull request by dagar**
- use full 20 bit data (18 bit accel, 19 bit gyro)
 - increase ODR to 16 kHz
 - disable all filters (anti-aliasing, notch filter, and UI filter all disabled or minimized for now until we want to set them intentionally)

Screenshot from 2020-08-30 21-55-26

The 20 bit data format is a bit weird (see below). To handle this and not grow our existing 16 bit sensor_accel_fifo/sensor_gyro_fifo messages I truncate the least significant bits as needed with a corresponding change to the scale factor. For example, since accelerometer data is is 18 bit dynamic range 16G it can fit within int16_t up until 4Gs.

Screenshot from 2020-08-30 21-54-09

### CODE REVIEW

**Comment by bkueng** [comment]
I gave this a quick spin in hover, and it performs as expected. The only thing I noticed is that CPU load is quite a bit higher, ~4% (compared to a slightly older version on master).

**Comment by dagar** [comment]
> I gave this a quick spin in hover, and it performs as expected. 

Thanks for the flight test, I've only tried this on the test rack. Was it the primary gyro? Do you happen to know if you crossed any of the thresholds that changed the scaling (sensor_{accel,gyro}_fifo.scale)? 

> The only thing I noticed is that CPU load is quite a bit higher, ~4% (compared to a slightly older version on master).

That's mostly the change to 16 kHz, which is almost certainly overkill, but I thought might be interesting for analysis until we can make a more informed decision. It also fit within our existing FIFO messages at typical loop rates and seemed like an increase we could afford on the newer boards that were likely to have this sensor (stm32f7, h7, etc). All that is to say, we can lower it to 8 kHz if you'd rather have the few percent cpu.

**Comment by dagar** [comment]
The other reason I thought the higher rate might be interesting is with the onboard FFT (https://github.com/PX4/Firmware/pull/15104) to get more granularity in determining frequency peaks and then ultimately set narrower notch filters.

File: src/drivers/imu/invensense/icm42688p/ICM42688P.cpp [hunk 1]
```diff
@@ -589,24 +561,100 @@ void ICM42688P::FIFOReset()
...
+	// 0xXXXAABBC
+	uint32_t high   = ((a << 12) & 0x000FF000);
+	uint32_t low    = ((b << 4)  & 0x00000FF0);
+	uint32_t lowest = (c         & 0x0000000F);
+
+	uint32_t x = high + low + lowest;
+
+	if (a & Bit7) {
+		// sign extend
+		x = 0xFFF00000u + x;
```
Comment (bkueng) [inline src/drivers/imu/invensense/icm42688p/ICM42688P.cpp:575]: Using bit ops might be a bit faster:
```suggestion
	uint32_t x = high | low | lowest;

	if (a & Bit7) {
		// sign extend
		x |= 0xFFF00000u;
```

**Review by bkueng — COMMENTED** [review]
> Was it the primary gyro? Do you happen to know if you crossed any of the thresholds that changed the scaling (sensor_{accel,gyro}_fifo.scale)?

Yes primary, but I didn't check the scaling.

> That's mostly the change to 16 kHz, which is almost certainly overkill, but I thought might be interesting for analysis until we can make a more informed decision. It also fit within our existing FIFO messages at typical loop rates and seemed like an increase we could afford on the newer boards that were likely to have this sensor (stm32f7, h7, etc). All that is to say, we can lower it to 8 kHz if you'd rather have the few percent cpu.

It should be fine but I have a feeling we'll hit the limit with certain configurations on Skynode. We can come back to it though if needed.

**Pull request merged**